AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -703,6 +704,89 @@ - (instancetype)styledWithBlock:(void (^)(ASLayoutElementStyle *style))styleBloc
 
 #pragma mark - Layout
 
+- (void)invalidateSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+
+  // Mark the node for layout in the next layout pass
+  [self invalidateCalculatedLayout];
+  
+  if (_supernode) {
+    ASDisplayNode *supernode = _supernode;
+    __instanceLock__.unlock();
+    // Cause supernode's layout to be invalidated
+    // We need to release the lock to prevent a deadlock
+    [supernode invalidateSize];
+    return;
+  }
+  
+  // Calculate a new pending layout. It will be applied in the next layout pass
+  ASLayout *layout = [self layoutThatFits:_calculatedDisplayNodeLayout->constrainedSize];
+  if (CGSizeEqualToSize(self.bounds.size, layout.size) == NO) {
+    // If the size of the layout changes inform the node of this
+    [self displayNodeDidInvalidateSizeOldSize:self.bounds.size];
+  }
+  
+  
+  __instanceLock__.unlock();
+}
+
+- (void)displayNodeDidInvalidateSizeOldSize:(CGSize)size
+{
+  // The default implementation of display node changes the size of itself to the new size
+  CGRect oldBounds = self.bounds;
+  CGSize oldSize = oldBounds.size;
+  CGSize newSize = _calculatedDisplayNodeLayout->layout.size;
```

Comment:
We should consider creating a method that can provide the best known constrainedSize at any point. It would check pending layout, current layout, and fallback to bounds if necessary. I think there is at least one other place that we use this, so sharing the logic would be nice.